### ISSUE TITLE
"Test tween command" and inbetween pop-up [$200]

### ISSUE THREAD

**Issue by ghost** [opentoonz/opentoonz#1423]
It 's a request to save time in the tweening process workflow. (already posted on the forum https://groups.google.com/forum/#!category-topic/opentoonz_en/feature-requests/JFhxb9NvPeo)

Currently in opentoonz the step to create a tweening is done by selecting the two keys in a range after adding empty inbetween frames by insert command.

It would be a huge time saver if we could preview this step in a single command (for exemple available on xsheet toolbar)

a button to simulate the process to create a tweening (it looks like a script ?) and use it as corrector 

without having to do it manually as long as we are not on a finalization step, by drawing only the keys from beginning.

The goal is to be able to check if our strokes/drawing goes in the right direction without finishing it completely.

The recent guided drawing feature is a great addition, and I think combining with this feature, we can have a total control in this type of process.

Ideally, we should preview with any key in the level strip or xsheet.

And select in a second button the number of inbetweening frames taken into account something between 1 and 12 for example. I can give more example if necessary.

https://1drv.ms/i/s!AspsYffPSwbdh9wtEIBJK5JDc4hDTw

The idea comes from cacani  "Quickflip" : https://www.youtube.com/watch?v=1LIrPyaG03o

I don't know the feasibility rate but opentoonz has many elements in common to do it.

If other people are interested please help me to promote it

PS : Sorry about the bad English I hope this is understandable, 

I will submit this to bountysource

**Comment by openanim** [comment]
I create the bounty here : https://www.bountysource.com/issues/48479745-check-strokes-command

Note: Take into consideration with this bounty that I have no idea how difficult it is to implant this.

**Comment by morevnaproject** [comment]
Hello @openanim! Thank you very much for initiative and providing a bounty! I will do my best to spread a word about it. ^__^

**Comment by openanim** [comment]
thanks @morevnaproject , I appreciate it :)

**Comment by blurymind** [comment]
This would be a huge time saver for me! One thing to note is setting the number of inbetweens and the type of tweening needs to be set somewhere for the quickflip preview - somewhere convenient.
Cacani has the advantage of having a dedicated gui element for onion skin settings and another one for tweening
![quickflip](https://user-images.githubusercontent.com/6495061/29961040-fc826080-8ef5-11e7-8fa4-de3b1ffd1bb8.PNG)
Another feature that would make this feature absolutely priceless is if had a quick command to generate a new inbetween from the quickflip after we are happy with it.

You could rename this post to something that notes that it is a feature request to improve the tweening workflow - "Test tween command" or something like that. "check strokes" could also suggest a feature for checking strokes while using the bucket tool.

**Comment by blurymind** [comment]
You could use the already existing tweening pop up that is used to generate inbetweens to set up the interpolation and duration of the preview tween ? But you would need to add an input box to it- to set the number of preview inbetweens. 

An improvement I would like to see is having the inbetween button available above the levels thumbnails. Having the ability to preview and create inbetweens by just selecting 2 frames and having OT create empty frames. 

So instead of this:
![tweencurrent](https://user-images.githubusercontent.com/6495061/29962214-4efce55c-8efa-11e7-8b72-d61c3cde92d6.png)
- you have to manually create empty frames before making a tween. It is annoyingly hard to reduce the number of inbetweens, which is a common operation when playing with timing
- You can only use the tween popup to generate tweens, it can not be used to set a tween preview requested here

We can have this:
![bettertweeninggui-mockup](https://user-images.githubusercontent.com/6495061/29962037-87eef7e8-8ef9-11e7-832e-14f0432afa69.png)
 In
- Instead of making empty frames manually, we could have the inbetween command only take the first and last frame and create the empty frames inbetween automatically and populate them. 
- The inbetween button could be always visible and enabled whenever the user has selected two frames
-The inbetween popup can be used to also set the inbetween preview requested here. The preview itself can be triggered via a keyboard shortcut of course

**Comment by beeheemooth** [comment]
Personally, I calculate in advance how many inbetweens will be needed, guided by the level with rough animation.
 The most uncomfortable thing in the entire process of creating automatic inbetveening is the compulsion to click to the right of the strip. In my opinion, this operation should be able to be called using hotkey. Since in OpenToonz interpolation is not configured as flexibly as in Cacani, it can simply call all four modes with four hotkeys: _Linear, Ease In, Ease Out, Ease In / Ease Out_ avoiding at the same pop-up windows.

**Comment by openanim** [comment]
@blurymind 

Thanks for suggesting a new title to get more relevant, you explained things better, than I did. That's exactly my thinking just like your additions that would really bring usability. This panel could be perfet. Thank you for clarifying and bringing concrete elements. I think it would be a great time saver in this kind of process. Personally, I use the function every 15 seconds or less in cacani, depending on the complexity of the drawing.

**Comment by openanim** [comment]
@beeheemooth 

About flexibility let's try to improve it to get closer to what is better or to do better in this matter.

**Comment by artisteacher** [comment]
@blurymind's suggestion for the inbetween pop-up looks super helpful. If these options were added, a single hotkey to access the pop-up would be great.

**Comment by openanim** [comment]
I added a reference to the title about it

**Comment by blurymind** [comment]
Thank you. Hope it gets an interested developer. This will get funded in no time imo 👍

**Comment by blurymind** [comment]
@artisteacher my idea for this is to have  the pop up called to setup the number of frames. In order to preview you could either click on the preview button or have a keyboard shortcut for previewing.  

I would like to re-address my proposal with another idea. Why don't we tie this with @turtletooth  's excellent onion skin mode - vector guided mode?

So instead of having to select any frames in the level strip to define what the first and last drawing are for the tween preview and generation, we just use what vector guided mode is using, like this:

![vectorguidedtweenpreview](https://user-images.githubusercontent.com/6495061/29996000-3a91e44c-8fee-11e7-8ad7-6d5cf927cb15.png)

**Comment by artisteacher** [comment]
I think the pop-up should also take into consideration workflows where the frames are already created. Ie. Have an additional “overwrite” or “use selected frames” option. When checked, this might disable the frame number field and use the selected drawings instead of creating new ones.

**Comment by RodneyBaker** [comment]
This feature appears to be slipping to 2019 development cycle.
Tagged with 'roadmap' due to active bounty.

**Comment by RodneyBaker** [comment]
No update in a years time although there is still a bounty.
Unfortunately it looks like the bounty has dropped from #$245 to $200.

Blurymind said:

> ...This will get funded in no time...

Unfortunately, after two years interest no developer has expressed interest (although any bounty against a feature is a good thing!)  Perhaps we need a review of this project to see what it requires to draw more attention.  Otherwise it might be advisable for supporters to pool their bounties on other projects to get them over the finish line... then return the focus to this feature at a later time.

**Comment by RodneyBaker** [comment]
@openanim  It will be good to review this request in light of the new inbetweening capability provided by #2931, specifically with interest in what shortfalls still exist in the auto-inbetweening workflow.

When reading this request the primary 'feature' of the request appeared to be the previewing capability.  A question then might be; to what extent does the vector guided drawing auto-inbetweening workflow (via Onionskinning) provide this capability?

**Comment by RodneyBaker** [comment]
I see a few possible ways forward on this request (relative to bounty):

- Award bounty if it is determined the auto inbetweening of guided drawing via #2021 meets the criteria for award
-- If yes, the bounty can be released and this feature request closed
-- If no, identifying what additional changes/updates to #2021 to have it achieve the goal would be optimal

- Maintain bounty for 'Test Tween Command' and reiterate what requirements need to be met for the feature/feature set.

- Transfer bounty to other target as appropriate.

- Something I've not identified above. 

I must assume the goal is to see this effort move forward so @openanim...  over to you!  :)

**Comment by openanim** [comment]
@RodneyBaker 

> When reading this request the primary 'feature' of the request appears to be the previewing capability. A question then might be; to what extent does the vector guided drawing auto-inbetweening workflow (via Onionskinning) provide this capability?

The two concepts are not really the same. In cacani this gives a preview of the previous key with the next key by generating temporary tweens. But we are not necessarily in production, it's just for checking.

https://www.youtube.com/watch?v=SSez0NFX834

https://www.youtube.com/watch?v=1LIrPyaG03o

You can decide the number of frames to take into account in the preview and the playback speed with frame rate option.

The new implementation is a mix of frame range and GD, compared to demand, we are directly producing tweens. They are not temporary.

In addition, if you create 2 keys one below the other, there will be no indication since there is no empty frame. The problem is that OT does not have a real key system.

Consequently, this has an impact on another point, with the current system the drawings do not refresh any changes  even with auto -inbetween activated. If you make a modification on one of your supposed key drawings, the in-between will not be updated.

To sum up :

- Preview between 2 drawings keys temporarily

- Have an automatic refresh option

- choose framed into account  and synchronysed with FPS

- This may require a review of the key system in OT. Or take inspiration from the blurymind proposal based on onion skins.

- Shortcut

But the great work of @manongjohn is not so far away, there is surely something to dig.

> Maintain bounty for 'Test Tween Command' and reiterate what requirements need to be met for the feature/feature set.
> 
> Transfer bounty to other target as appropriate.
> 
> Something I've not identified above.
> 
That does not bother me. Even if it's @Orphanlast to decide that. In the past, I had already advised him to change premiums if necessary, in order to avoid his money sleeping unnecessarily.

**Comment by RodneyBaker** [comment]
So in the interest of clarityfor where the proposed bounty will be awarded I see the following possible milestones:
Milestones
1.  Create a temporary preview between two (2) tweenable key drawings
2.  Include an automatic refresh option (to refresh the temporary preview when a key drawing is altered)
3.  ***choose framed into account and synchronized with FPS*** _Not quite sure how to interpret this one_  I assume this means to take all frames into account allowing for autoinbetweening between any number of 'empty' frames between two (2) keyframes based on current FPS. 
4.  Include a Shortcut/hotkey for activation of the feature.  _specifics not cited_

The above feature is expected to entail the following general requirements :
A.  Conduct a review of the current key system in OT.   
B.  Updating of the key system may be required but is not explicitly a requirement for award of this bounty although reward may be delayed until such a time that the key system can support implementation of the desired feature.

Additional reference:  
Please take inspiration from the @blurymind proposal based on onion skins.  _see post by @blurymind for details:  [LINK](https://github.com/opentoonz/opentoonz/issues/1423#issuecomment-326526898)_

Disclaimer:  The above is not intended to be any official statement of award criteria for this or any other bounty associated with Opentoonz development but is more toward trying to better understand the currently utilized process associated with Opentoonz bounties.



Aside:  After tax season is over I hope to test the waters on a bounty or two to see how we can best/better utilize Bountysource while avoiding the common delays experienced in the process. 
 Bounty/incentives should be awarded quickly to the greatest extent possible and providing clear and specific targets/milestones will help in this regard.  Of late @Orphanlast hasn't been very active in the OT community so recommend each contributor take as much ownership of their contributions/bounties as reasonably possible.

**Comment by openanim** [comment]
about 3/ I misspoke, it's just the number of frames used in the preview. In cacani this goes up to 10 frames taken into account as an example. And you can choose the frame rate you want applied on it. I don't think that on this side will be a problem.

**Comment by openanim** [comment]
@Orphanlast

Following the modification of the terms of bountysource (see the last comment in # 1349).

I think you should take care of this bounty to put it on another issue or claim your money before the deadline.

**Comment by morevnaproject** [comment]
I've got a mail that they decided to withdraw the ridiculous changes - 
![screenshot_004](https://user-images.githubusercontent.com/332868/85253338-248d1880-b488-11ea-9144-2e57d0ac5bfa.png)

**Comment by b4zz4** [comment]
I'm working on my own version of self-guided animation based on ARAP's ideas for now it's pixelating. But vector points can also be distorted using FLOW

Although deepflow is a bit slow, this idea can be improved and moved to opentoonz

![vejete](https://user-images.githubusercontent.com/643833/102634652-e0c25580-4130-11eb-999f-062e5969a1b6.gif)

https://4232.cf/es/automatic-animation-based-deepflow-ebsynth/

**Comment by RodneyBaker** [comment]
@openanim 
Given that five years have passed and no developer has shown interest in working on this feature I will recommend reassigning the bounty.   There are other options of course such as trying to get the word out to increase the bounty and to strir up general interest in the feature itself.  

Bottom line from my perspective is that bounties should not be held for long term as that defeats the purpose of a bounty.  If there isn't interest then I would consider other areas of interest to apply the bounty.

In the end this is your call but it has become rather obvious (to me) that this particular bounty system has not been working as original anticipated.

**Comment by openanim** [comment]
@RodneyBaker Personally, I don't have a problem with that. I had already talked to @Orphanlast in the past about putting it on another issues. But there was no indication from him. As long as I can remember.

**Comment by openanim** [comment]
Among the issues that could receive the transfer of this bounty. I was personally thinking about the problems of importing and exporting the svg. What do you think? Of course it's just a suggestion. @RodneyBaker For your information, I contacted @Orphanlast to find out what he recommended in this regard.
